packages/experience-builder-sdk/src/blocks/editor/VisualEditorContext.tsx

```diff
@@ -78,14 +83,43 @@ export function VisualEditorContextProvider({
   const [isDragging, setIsDragging] = useState(false);
   const selectedNodeId = useRef<string>('');
   const [locale, setLocale] = useState<string>(initialLocale);
+  const [areEntitiesFetched, setEntitiesFetched] = useState(false);
 
   const entityStore = useRef<EditorModeEntityStore>(
     new EditorModeEntityStore({
-      entities: [],
+      // Initially, the SDK boots in preview mode and already fetches entities.
+      // We utilizes the previosuly existing store to avoid loading twice and
+      // have the entities ready as early as possible.
+      entities: previousEntityStore?.entities ?? [],
       locale: locale,
     })
   );
 
+  // Reload the entity store when the locale changed
+  useEffect(() => {
+    if (!locale || locale === entityStore.current.locale) return;
+    entityStore.current = new EditorModeEntityStore({
+      entities: [],
+      locale: locale,
+    });
+  }, [locale]);
+
+  // When the tree was updated, we store the dataSource and
+  // afterward, this effect fetches the respective entities.
+  useEffect(() => {
```

Comment:
Was there a reason for moving this useEffect in the context? Thinking about the issues we were seeing with contexts triggering unnecessary network requests, isn't it better for this useEffect to stay in the component so it doesn't accidentally trigger every time this context is used?

Reply:
Technically, this doesn't change anything. The `<VisualEditorContextProvider>` is only used once in the wrapper of `<VisualEditorRoot>`. So on every render cycle, the logic will once run through the context component **and** once through the root component. It's triggered exactly the same amount of times as before.

Note that `useContext` doesn't render any effects but just returns the generated object that is returned by the provider. So, consuming the context values via `useContext` is not the same as "using"/ "rendering" the context multiple times.

Reply:
You right, the problem we've seen in user_interface was not related to `useContext` but regular hooks that included `useEffect`